Referral program implemented #295

maslenitsa93 opened the pull request:
#295 

# Creating referral account. Hardfork check. Disabled breaking referral

1. Create `test` account and it's `testref` referral account.

```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
create_account cyberfounder test "{}" "300.000 GOLOS" true
transfer_to_vesting cyberfounder test "30.000 GOLOS" true
transfer cyberfounder test "0.200 GOLOS" "" true
create_account_referral test "0.200 GOLOS" "0.000001 GESTS" testref "{}" {"referrer": "test", "interest_rate": 900, "end_date": "2018-09-26T14:00:00", "break_fee": "0.000 GOLOS"} true
```

2. You have a fail with message about unsupported operation because you done it before 19 hardfork applied. Wait for this hardfork and run again.

3. See the Mongo.

Result: account_object has an object will following fields:
```
   ...
    "referrer_account" : "test", 
    "referrer_interest_rate" : NumberLong(900), 
    "referral_end_date" : ISODate("2018-09-26T14:00:00.000+0000"), 
    "referral_break_fee_value" : 0.0, 
    "referral_break_fee_symbol" : "GOLOS", 
    ...
}
```

4. See the `get_accounts` API with following HTML page in browser:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.getAccounts(['test', 'testref'], function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result:
```
    "referrer_account": "test",
    "referrer_interest_rate": 900,
    "referral_end_date": "2018-09-26T14:00:00",
    "referral_break_fee": "0.000 GOLOS"
  }
]
```

5. Try to break referral program:
```
break_free_referral testref true
```
Result: fail - no right to break referral.

# Enabled breaking referral. Try to break referral without enough balance

1. Create another referral account:
```
create_account_referral test "0.200 GOLOS" "0.000001 GESTS" testref2 "{}" {"referrer": "test", "interest_rate": 900, "end_date": "2018-09-26T14:00:00", "break_fee": "0.001 GOLOS"} true
```

2. Check it:
```
get_account testref2
```
Result:
```
  ...
  "referrer_account": "test",
  "referrer_interest_rate": 900,
  "referral_end_date": "2018-09-26T14:00:00",
  "referral_break_fee": "0.001 GOLOS"
}
```

3. Try to break referral (without enough balance):
```
break_free_referral testref2 true
```
Result: fail - does not have sufficient funds.

4. Transfer some money to balance and try again:
```
transfer cyberfounder testref2 "0.200 GOLOS" "" true
break_free_referral testref2 true
```
Result: success.

5. Check referral (in past) account:
```
get_account testref2
```
Result:
```
  ...
  "balance": "0.199 GOLOS",
  ...
  "referrer_account": "",
  "referrer_interest_rate": 0,
  "referral_end_date": "1970-01-01T00:00:00",
  "referral_break_fee": "0.000 GOLOS"
}
```

6. Check referrer account:
```
get_account test
```
Result:
```
  ...
  "balance": "0.001 GOLOS",
  ...
}
```

# Creating comment before breaking referral and after it

1. Create referral account:
```
create_account_referral test "0.200 GOLOS" "0.000001 GESTS" testref3 "{}" {"referrer": "test", "interest_rate": 900, "end_date": "2018-09-26T14:00:00", "break_fee": "0.001 GOLOS"} true
```

2. Post comment from it:
```
post_comment testref3 test "" test hello world "{}" true
```

3. Check comment in Mongo. Result:
```
    ...
    "beneficiaries" : [
        {
            "account" : "test", 
            "weight" : NumberLong(900)
        }
    ], 
    ...
```

4. Break referral:
```
transfer cyberfounder testref3 "0.200 GOLOS" "" true
break_free_referral testref3 true
```

5. Wait for 5 minutes and post comment again:
```
post_comment testref3 test2 "" test hello world "{}" true
```

6. Check comment in Mongo. 
Result: no object with account="test" in `beneficiaries`.

maslenitsa93 commented:
Do not write me about code style in code what is not my but is near. I will do big code style fix after all work.

afalaleev commented:
1.  should be enough one of the:
- `std::find_if` 
-  checking in the loop https://github.com/GolosChain/golos/blob/golos-v0.19.0/libraries/chain/steem_evaluator.cpp#L472-L475
2. in the `comment_options_evaluator` you should check 
- `beneficiaries.size() == 0 || (beneficiaries.size() == 1 && beneficiaries.front() == referrer)`
- `sum of weights <= 100%`
- `number of beneficiaries` 
- `unique names of beneficiaries`

maslenitsa93 commented:
Bug fixed:
https://github.com/GolosChain/golos/pull/942/commits/b29cbc8ad21f714e91d2e7064be6ddbe5b19c95b#diff-ddf0678389d8cc985093894138e0c595L141

And this one couldn't be fixed:
https://github.com/GolosChain/golos/pull/942/commits/b29cbc8ad21f714e91d2e7064be6ddbe5b19c95b#diff-ddf0678389d8cc985093894138e0c595L151
because it is not good to change messages of exceptions (parsing of them is dirty but it don't stop some persons)

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -233,6 +233,42 @@ namespace golos { namespace chain {
+            void operator()(const account_referral_options& aro) const {
+                if (!_db.has_hardfork(STEEMIT_HARDFORK_0_19__295)) {
+                    FC_THROW_EXCEPTION(golos::unsupported_operation,
+                        "account_referral_options is not supported until STEEMIT_HARDFORK_0_19__295");
```
Replace with `ASSERT_REQ_HF()`

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -656,14 +722,26 @@ namespace golos { namespace chain {
                             com.cashout_time = com.created + STEEMIT_CASHOUT_WINDOW_SECONDS;
                         }
 
+                        if (_db.has_hardfork(STEEMIT_HARDFORK_0_19__295)
```
Not needed case.
Referrer appears on create of account, it can't appear after some time.
So, additing of beneficiary can happen only on create post/comment.

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -2288,4 +2366,30 @@ namespace golos { namespace chain {
+            if (!_db.has_hardfork(STEEMIT_HARDFORK_0_19__295)) {
+                FC_THROW_EXCEPTION(golos::unsupported_operation,
+                    "break_free_referral_operation is not supported until STEEMIT_HARDFORK_0_19__295");
+            }
```
`ASSERT_REQ_HF()`

maslenitsa93 commented:
ASSERT_REQ_HF:

- Using of `db()` makes this macro unworkable in many cases including my one. _db seems better. Replaced.
- But additionally, added `ASSERT_REQ_HF_IN_DB` macro for cases with custom db variable.

afalaleev commented on libraries/wallet/wallet.cpp:
```diff
@@ -2181,23 +2240,29 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
+            SET_PROP(p, create_account_delegation_time);
+            SET_PROP(p, min_delegation);
+            op.props = p;
+            if (!!props.max_referral_interest_rate || !!props.max_referral_term_sec
```
The logic of `wallet_api::update_chain_properties()` is focused on replace only changed values.
You break it. 
An example of broken logic:
1. User set `max_referral_interest_rate` with one call of cli_wallet
2. After some time he decide to change `account_creation_fee`
3. cli_wallet will send `chain_properties_18` which flushes `max_referral_interest_rate` back to default value.

You should check hardfork version of blockchain and use suitable version of chain_properties.

maslenitsa93 commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -656,14 +719,25 @@ namespace golos { namespace chain {
+                                    com.beneficiaries.end(), [&auth](const beneficiary_route_type& benef) {
+                                return benef.account == auth.referrer_account;
+                            });
+                            if (referrer_find == com.beneficiaries.end()) {
```
@afalaleev "Where is case with clearing of referrer_account?" What is clearing do you want? Removing referrer on editing of comment if referral breaked?

afalaleev replied:
Yes.

afalaleev replied:
And add of clearing referrer in API if date is end.

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -445,35 +482,57 @@ namespace golos { namespace chain {
+                        && _c.beneficiaries.front().account == _a.referrer_account) {
+                    total_weight += _c.beneficiaries[0].weight;
+                    auto& referrer = _a.referrer_account;
+                    const auto& referrer_find = std::find_if(cpb.beneficiaries.begin(), cpb.beneficiaries.end(),
```
Keep current naming rules - rename to `itr`.

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -2288,4 +2367,27 @@ namespace golos { namespace chain {
+            const auto& referrer = _db.get_account(referral.referrer_account);
+
+            if (referral.referral_break_fee.amount == 0) {
+                FC_THROW_EXCEPTION(golos::unsupported_operation,
```
`logic_exception::no_right_to_break_referral` ?

afalaleev commented on libraries/protocol/include/golos/protocol/steem_operations.hpp:
```diff
@@ -25,6 +25,21 @@ namespace golos { namespace protocol {
+            void validate() const;
+        };
+
+        typedef static_variant <
```
using

afalaleev commented on libraries/protocol/include/golos/protocol/steem_operations.hpp:
```diff
@@ -25,6 +25,21 @@ namespace golos { namespace protocol {
+            account_referral_options
+        > account_create_with_delegation_extension;
+
+        typedef flat_set<account_create_with_delegation_extension> account_create_with_delegation_extensions_type;
```
using

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -445,35 +482,57 @@ namespace golos { namespace chain {
                     }
                 });
+
+                GOLOS_CHECK_LOGIC(total_weight <= STEEMIT_100_PERCENT,
```
Now, `comment_payout_beneficiaries::validate()` and `comment_options_extension_visitor::operator()` returns different errors on same cases.
Errors should be the same.

afalaleev commented on libraries/chain/steem_evaluator.cpp:
```diff
@@ -445,35 +482,56 @@ namespace golos { namespace chain {
                     }
                 });
+
+                GOLOS_CHECK_VALUE(total_weight <= STEEMIT_100_PERCENT,
```
Errors still diffirent.
